File: apps/server/internal/edge/canonical.go
package edge

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"math"
	"sort"
	"strconv"
	"strings"
	"unicode/utf16"
)

// Canonical JSON for signed Edge bodies: the RFC 8785 (JCS) subset shared
// with edge_protocol::canonical in the Rust daemon. Values are null,
// booleans, strings, arrays, objects and integers in the IEEE-754 safe range;
// any other number is rejected so a signature never depends on float
// formatting. Object members are sorted by UTF-16 code units and strings are
// escaped exactly as ECMAScript JSON.stringify does. Both implementations are
// held to packages/edge-protocol/fixtures/canonical-json.

const maxSafeInteger = 1<<53 - 1

const maxCanonicalDepth = 16

var (
	ErrUnsupportedNumber = errors.New("signed JSON may contain only integers in the safe range")
	ErrCanonicalDepth    = errors.New("signed JSON nesting is too deep")
	ErrNotCanonical      = errors.New("body is not in canonical form")
)

// Canonicalize serializes a value built from map[string]any, []any, string,
// bool, nil, json.Number and Go integer types.
func Canonicalize(value any) ([]byte, error) {
	var out bytes.Buffer
	if err := writeCanonical(&out, value, 0); err != nil {
		return nil, err
	}
	return out.Bytes(), nil
}

// CanonicalizeJSON parses JSON text (preserving number text) and returns its
// canonical form.
func CanonicalizeJSON(text []byte) ([]byte, error) {
	decoder := json.NewDecoder(bytes.NewReader(text))
	decoder.UseNumber()
	var value any
	if err := decoder.Decode(&value); err != nil {
		return nil, fmt.Errorf("invalid JSON: %w", err)
	}
	if decoder.More() {
		return nil, errors.New("invalid JSON: trailing data")
	}
	return Canonicalize(value)
}

// ParseCanonical decodes text and requires it to already be canonical,
// rejecting duplicate keys, whitespace and escape variants.
func ParseCanonical(text []byte) (map[string]any, error) {
	again, err := CanonicalizeJSON(text)
	if err != nil {
		return nil, err
	}
	if !bytes.Equal(again, text) {
		return nil, ErrNotCanonical
	}
	decoder := json.NewDecoder(bytes.NewReader(text))
	decoder.UseNumber()
	var value map[string]any
	if err := decoder.Decode(&value); err != nil {
		return nil, err
	}
	return value, nil
}

func writeCanonical(out *bytes.Buffer, value any, depth int) error {
	if depth > maxCanonicalDepth {
		return ErrCanonicalDepth
	}
	switch v := value.(type) {
	case nil:
		out.WriteString("null")
	case bool:
		if v {
			out.WriteString("true")
		} else {
			out.WriteString("false")
		}
	case string:
		writeCanonicalString(out, v)
	case json.Number:
		text := v.String()
		if strings.ContainsAny(text, ".eE") {
			return ErrUnsupportedNumber
		}
		n, err := strconv.ParseInt(text, 10, 64)
		if err != nil || n > maxSafeInteger || n < -maxSafeInteger {
			return ErrUnsupportedNumber
		}
		out.WriteString(strconv.FormatInt(n, 10))
	case int:
		return writeCanonical(out, int64(v), depth)
	case int32:
		return writeCanonical(out, int64(v), depth)
	case int64:
		if v > maxSafeInteger || v < -maxSafeInteger {
			return ErrUnsupportedNumber
		}
		out.WriteString(strconv.FormatInt(v, 10))
	case uint32:
		return writeCanonical(out, int64(v), depth)
	case uint64:
		if v > maxSafeInteger {
			return ErrUnsupportedNumber
		}
		out.WriteString(strconv.FormatUint(v, 10))
	case float64, float32:
		f := toFloat(v)
		if math.Trunc(f) != f {
			return ErrUnsupportedNumber
		}
		// Integers decoded without UseNumber arrive as float64; accept only
		// exact values in range.
		if f > maxSafeInteger || f < -maxSafeInteger {
			return ErrUnsupportedNumber
		}
		out.WriteString(strconv.FormatInt(int64(f), 10))
	case []any:
		out.WriteByte('[')
		for i, item := range v {
			if i > 0 {
				out.WriteByte(',')
			}
			if err := writeCanonical(out, item, depth+1); err != nil {
				return err
			}
		}
		out.WriteByte(']')
	case map[string]any:
		keys := make([]string, 0, len(v))
		for key := range v {
			keys = append(keys, key)
		}
		sort.Slice(keys, func(i, j int) bool { return lessUTF16(keys[i], keys[j]) })
		out.WriteByte('{')
		for i, key := range keys {
			if i > 0 {
				out.WriteByte(',')
			}
			writeCanonicalString(out, key)
			out.WriteByte(':')
			if err := writeCanonical(out, v[key], depth+1); err != nil {
				return err
			}
		}
		out.WriteByte('}')
	default:
		return fmt.Errorf("unsupported canonical JSON value %T", value)
	}
	return nil
}

func toFloat(v any) float64 {
	switch f := v.(type) {
	case float32:
		return float64(f)
	case float64:
		return f
	}
	return math.NaN()
}

func lessUTF16(a, b string) bool {
	ua, ub := utf16.Encode([]rune(a)), utf16.Encode([]rune(b))
	for i := 0; i < len(ua) && i < len(ub); i++ {
		if ua[i] != ub[i] {
			return ua[i] < ub[i]
		}
	}
	return len(ua) < len(ub)
}

func writeCanonicalString(out *bytes.Buffer, value string) {
	const hex = "0123456789abcdef"
	out.WriteByte('"')
	for _, r := range value {
		switch r {
		case '"':
			out.WriteString(`\"`)
		case '\\':
			out.WriteString(`\\`)
		case '\b':
			out.WriteString(`\b`)
		case '\f':
			out.WriteString(`\f`)
		case '\n':
			out.WriteString(`\n`)
		case '\r':
			out.WriteString(`\r`)
		case '\t':
			out.WriteString(`\t`)
		default:
			if r < 0x20 {
				out.WriteString(`\u00`)
				out.WriteByte(hex[r>>4])
				out.WriteByte(hex[r&0x0f])
			} else {
				out.WriteRune(r)
			}
		}
	}
	out.WriteByte('"')
}

File: apps/edge/crates/edge-protocol/src/canonical.rs
//! Canonical JSON for signed bodies (an RFC 8785 / JCS subset).
//!
//! Signed Tilecast Edge bodies are serialized with the JSON Canonicalization
//! Scheme, restricted to a subset that every implementation can reproduce
//! byte-for-byte without a floating-point formatter:
//!
//! * values: `null`, booleans, strings, arrays, objects, and **integers** in
//!   the IEEE-754 safe range `[-(2^53-1), 2^53-1]`. Any other number is
//!   rejected, so a signed body can never depend on float formatting.
//! * object members sorted by the UTF-16 code units of their names;
//! * no insignificant whitespace;
//! * strings escaped exactly as ECMAScript `JSON.stringify` does: `\"`, `\\`,
//!   `\b \f \n \r \t`, other controls as lowercase `\u00xx`, everything else
//!   emitted as literal UTF-8.
//!
//! Verification never re-serializes a typed struct. The verifier checks the
//! signature over the exact received bytes and then requires those bytes to
//! equal [`canonicalize`] of their own parse ([`parse_canonical`]). That makes
//! every accepted signed body the unique encoding of its value, which rules
//! out duplicate-key, whitespace and escape-variant tricks, and gives each
//! body one stable digest.
//!
//! The Go implementation lives in `apps/server/internal/edge/canonical.go`.
//! Both are held to `packages/edge-protocol/fixtures/canonical-json/`.

use serde_json::Value;

/// Largest integer magnitude a signed body may contain (2^53 - 1).
pub const MAX_SAFE_INTEGER: u64 = 9_007_199_254_740_991;
/// Nesting deeper than this is rejected; real bodies are shallow.
pub const MAX_DEPTH: usize = 16;

#[derive(Debug, Clone, PartialEq, Eq, thiserror::Error)]
pub enum CanonicalError {
    #[error("signed JSON may contain only integers in the safe range")]
    UnsupportedNumber,
    #[error("signed JSON nesting is too deep")]
    TooDeep,
    #[error("body is not valid JSON")]
    InvalidJson,
    #[error("body is not in canonical form")]
    NotCanonical,
}

/// Serializes `value` in canonical form.
pub fn canonicalize(value: &Value) -> Result<Vec<u8>, CanonicalError> {
    let mut out = Vec::with_capacity(256);
    write_value(value, &mut out, 0)?;
    Ok(out)
}

/// Parses `bytes` and requires them to already be canonical.
pub fn parse_canonical(bytes: &[u8]) -> Result<Value, CanonicalError> {
    let value: Value = serde_json::from_slice(bytes).map_err(|_| CanonicalError::InvalidJson)?;
    let again = canonicalize(&value)?;
    if again != bytes {
        return Err(CanonicalError::NotCanonical);
    }
    Ok(value)
}

fn write_value(value: &Value, out: &mut Vec<u8>, depth: usize) -> Result<(), CanonicalError> {
    if depth > MAX_DEPTH {
        return Err(CanonicalError::TooDeep);
    }
    match value {
        Value::Null => out.extend_from_slice(b"null"),
        Value::Bool(true) => out.extend_from_slice(b"true"),
        Value::Bool(false) => out.extend_from_slice(b"false"),
        Value::Number(number) => {
            if let Some(unsigned) = number.as_u64() {
                if unsigned > MAX_SAFE_INTEGER {
                    return Err(CanonicalError::UnsupportedNumber);
                }
                out.extend_from_slice(unsigned.to_string().as_bytes());
            } else if let Some(signed) = number.as_i64() {
                if signed.unsigned_abs() > MAX_SAFE_INTEGER {
                    return Err(CanonicalError::UnsupportedNumber);
                }
                out.extend_from_slice(signed.to_string().as_bytes());
            } else {
                return Err(CanonicalError::UnsupportedNumber);
            }
        }
        Value::String(text) => write_string(text, out),
        Value::Array(items) => {
            out.push(b'[');
            for (index, item) in items.iter().enumerate() {
                if index > 0 {
                    out.push(b',');
                }
                write_value(item, out, depth + 1)?;
            }
            out.push(b']');
        }
        Value::Object(members) => {
            let mut keys: Vec<&String> = members.keys().collect();
            keys.sort_by(|a, b| a.encode_utf16().cmp(b.encode_utf16()));
            out.push(b'{');
            for (index, key) in keys.into_iter().enumerate() {
                if index > 0 {
                    out.push(b',');
                }
                write_string(key, out);
                out.push(b':');
                write_value(&members[key.as_str()], out, depth + 1)?;
            }
            out.push(b'}');
        }
    }
    Ok(())
}

fn write_string(text: &str, out: &mut Vec<u8>) {
    const HEX: &[u8; 16] = b"0123456789abcdef";
    out.push(b'"');
    for character in text.chars() {
        match character {
            '"' => out.extend_from_slice(b"\\\""),
            '\\' => out.extend_from_slice(b"\\\\"),
            '\u{08}' => out.extend_from_slice(b"\\b"),
            '\u{0c}' => out.extend_from_slice(b"\\f"),
            '\n' => out.extend_from_slice(b"\\n"),
            '\r' => out.extend_from_slice(b"\\r"),
            '\t' => out.extend_from_slice(b"\\t"),
            c if (c as u32) < 0x20 => {
                let code = c as u32 as usize;
                out.extend_from_slice(b"\\u00");
                out.push(HEX[code >> 4]);
                out.push(HEX[code & 0x0f]);
            }
            c => {
                let mut buffer = [0u8; 4];
                out.extend_from_slice(c.encode_utf8(&mut buffer).as_bytes());
            }
        }
    }
    out.push(b'"');
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn canon(value: Value) -> String {
        String::from_utf8(canonicalize(&value).expect("canonical")).expect("utf8")
    }

    #[test]
    fn sorts_keys_by_utf16_code_units() {
        // U+1F600 (surrogate pair D83D DE00) sorts before U+FFFD in UTF-16
        // even though its UTF-8 bytes sort after.
        let value = json!({"\u{fffd}": 1, "\u{1f600}": 2, "b": 3, "a": 4, "A": 5});
        assert_eq!(canon(value), "{\"A\":5,\"a\":4,\"b\":3,\"\u{1f600}\":2,\"\u{fffd}\":1}");
    }

    #[test]
    fn escapes_like_json_stringify() {
        let value = json!("q\"b\\\u{8}\u{c}\n\r\t\u{1}\u{1f}\u{7f}é\u{2028}");
        assert_eq!(canon(value), "\"q\\\"b\\\\\\b\\f\\n\\r\\t\\u0001\\u001f\u{7f}é\u{2028}\"");
    }

    #[test]
    fn rejects_non_integers_and_unsafe_integers() {
        assert_eq!(canonicalize(&json!(1.5)), Err(CanonicalError::UnsupportedNumber));
        assert_eq!(canonicalize(&json!(MAX_SAFE_INTEGER + 1)), Err(CanonicalError::UnsupportedNumber));
        assert_eq!(canonicalize(&json!(-(MAX_SAFE_INTEGER as i64) - 1)), Err(CanonicalError::UnsupportedNumber));
        assert_eq!(canon(json!(-(MAX_SAFE_INTEGER as i64))), "-9007199254740991");
    }

    #[test]
    fn parse_canonical_rejects_variants() {
        assert!(parse_canonical(b"{\"a\":1,\"b\":[true,null]}").is_ok());
        for bad in [
            &b"{\"b\":1,\"a\":2}"[..],
            b"{\"a\": 1}",
            b"{\"a\":1,\"a\":1}",
            b"{\"a\":\"\\u0041\"}",
            b"{\"a\":1.0}",
            b"{\"a\":1e2}",
            b" {\"a\":1}",
        ] {
            assert!(parse_canonical(bad).is_err(), "{:?}", String::from_utf8_lossy(bad));
        }
    }

    #[test]
    fn depth_is_bounded() {
        let mut value = json!(0);
        for _ in 0..=MAX_DEPTH {
            value = json!([value]);
        }
        assert_eq!(canonicalize(&value), Err(CanonicalError::TooDeep));
    }
}
